core/corehttp/gateway_handler.go

```diff
@@ -308,12 +293,22 @@ func (i *gatewayHandler) getOrHeadHandler(ctx context.Context, w http.ResponseWr
 
 	// storage for directory listing
 	var dirListing []directoryItem
-	dirr.ForEachLink(ctx, func(link *ipld.Link) error {
+	dirit := dir.Entries()
+	for dirit.Next() {
 		// See comment above where originalUrlPath is declared.
-		di := directoryItem{humanize.Bytes(link.Size), link.Name, gopath.Join(originalUrlPath, link.Name)}
+		s, err := dirit.Node().Size()
+		if err != nil {
```

Comment:
We can *probably* just ignore this error.

Reply:
IIRC, filesize is technically optional (but I guess we don't have a spec yet).

Reply:
Link size is optional, I'm not sure about this one

Reply:
Digging deeper in will only fail if protobuf encoding fails - https://github.com/ipfs/go-merkledag/blob/master/node.go#L240 - at which point we probably want to give user the error

Reply:
"well documented": https://github.com/ipfs/go-ipfs/issues/4286#issuecomment-336195370

Reply:
This won't error on nodes which don't have size and encode successfully (note that it's not the unixfs filesize (which we don't set for directories, I think) - only the merkledag node size - https://github.com/ipfs/go-ipfs/pull/5649/files#diff-48d085152bf55180b23ed7d5bd1b9b0dR126)

Reply:
Is there anything else blocking this?